packages/components/src/components/form/bal-number-input/bal-number-input.tsx

```diff
@@ -0,0 +1,338 @@
+   */
+  @Event() balKeyPress!: EventEmitter<KeyboardEvent>
+
+  @Listen('click', { capture: true, target: 'document' })
+  listenOnClick(ev: UIEvent) {
+    if (this.disabled && ev.target && ev.target === this.el) {
+      ev.preventDefault()
+      ev.stopPropagation()
+    }
+  }
+
+  componentWillLoad() {
+    this.inheritedAttributes = inheritAttributes(this.el, ['aria-label', 'tabindex', 'title'])
+  }
+
+  componentDidLoad() {
+    this.didInit = true
+    if (!isNil(this.value)) {
+      this.valueChanged(this.value, undefined)
+    }
+  }
+
+  connectedCallback() {
+    this.debounceChanged()
+    attachComponentToConfig(this)
+  }
+
+  disconnectedCallback() {
+    detachComponentToConfig(this)
+  }
+
+  configChanged(state: BalConfigState): void {
+    this.language = state.language
+    this.region = state.region
+
+    if (!this.hasFocus && this.nativeInput) {
+      this.nativeInput.value = this.getFormattedValue()
+    }
+  }
+
+  /**
+   * Sets focus on the native `input` in `ion-input`. Use this method instead of the global
+   * `input.focus()`.
+   */
+  @Method()
+  async setFocus() {
+    if (this.nativeInput) {
+      this.nativeInput.focus()
+    }
+  }
+
+  /**
+   * Sets blur on the native `input` in `ion-input`. Use this method instead of the global
+   * `input.blur()`.
+   * @internal
+   */
+  @Method()
+  async setBlur() {
+    if (this.nativeInput) {
+      this.nativeInput.blur()
+    }
+  }
+
+  /**
+   * Returns the native `<input>` element used under the hood.
+   */
+  @Method()
+  getInputElement(): Promise<HTMLInputElement> {
+    return Promise.resolve(this.nativeInput!)
+  }
+
+  private getAllowedKeys() {
+    return [...NUMBER_KEYS, ...ACTION_KEYS, getDecimalSeparator()]
+  }
+
+  private getRawValue(): string {
+    return typeof this.value === 'number' ? this.value.toString() : (this.value || '').toString()
+  }
+
+  private getFormattedValue(): string {
+    const value = this.getRawValue()
+    const suffix = this.suffix !== undefined && value !== undefined && value !== '' ? ' ' + this.suffix : ''
+    return `${formatInputValue(value, this.decimal)}${suffix}`
+  }
+
+  private onInput = (ev: Event) => {
+    const input = ev.target as HTMLInputElement | null
+    if (input) {
+      const parsedValue = parseFloat(parseFloat(input.value).toFixed(this.decimal))
+      if (!isNaN(parsedValue)) {
+        this.value = parsedValue
+      } else {
+        if (this.decimal === 0 || input.value !== getDecimalSeparator()) {
+          this.value = null
+          input.value = ''
+        }
+      }
+    }
+    this.balInput.emit(this.value)
+  }
+
+  private onBlur = (ev: FocusEvent) => {
+    const input = ev.target as HTMLInputElement | null
+
+    this.hasFocus = false
+    this.balBlur.emit(ev)
+
+    if (input && input.value === getDecimalSeparator()) {
+      this.value = null
+      input.value = ''
+    }
+
+    if (this.cachedValue !== this.value) {
+      this.balChange.emit(this.value)
```

Comment:
shouldn't cached value be also updated here to the new value after emit?

Reply:
adjust change event